File: Medium/2025-11-06-3607-Power-Grid-Maintenance/solution.js
class UnionFind {
    constructor(size) {
        this.parent = Array.from({ length: size }, (_, i) => i);
        this.rank = Array(size).fill(1);
        this.online = Array(size).fill(true);
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX !== rootY) {
            if (this.rank[rootX] > this.rank[rootY]) {
                this.parent[rootY] = rootX;
            } else if (this.rank[rootX] < this.rank[rootY]) {
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX;
                this.rank[rootX]++;
            }
        }
    }

    setOffline(x) {
        this.online[x] = false;
    }

    getMinOnlineInComponent(x) {
        const root = this.find(x);
        let minId = -1;
        for (let i = 0; i < this.online.length; i++) {
            if (this.find(i) === root && this.online[i]) {
                minId = minId === -1 ? i : Math.min(minId, i);
            }
        }
        return minId + 1; // Convert to 1-based index
    }
}

var maintenanceCheck = function(c, connections, queries) {
    const uf = new UnionFind(c);
    for (const [u, v] of connections) {
        uf.union(u - 1, v - 1);
    }
    
    const result = [];
    for (const [type, x] of queries) {
        const station = x - 1; // Convert to 0-based index
        if (type === 1) {
            const minOnline = uf.getMinOnlineInComponent(station);
            result.push(minOnline);
        } else if (type === 2) {
            uf.setOffline(station);
        }
    }
    
    return result;
};

File: Medium/2025-11-06-3607-Power-Grid-Maintenance/solution.py
class UnionFind:
    def __init__(self, size):
        self.parent = list(range(size))
        self.rank = [1] * size
        self.online = [True] * size

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1

    def set_offline(self, x):
        self.online[x] = False

    def get_min_online(self, x):
        root = self.find(x)
        min_online = float('inf')
        for i in range(len(self.online)):
            if self.find(i) == root and self.online[i]:
                min_online = min(min_online, i)
        return min_online if min_online != float('inf') else -1


class Solution:
    def maintenance(self, c: int, connections: List[List[int]], queries: List[List[int]]) -> List[int]:
        uf = UnionFind(c + 1)
        
        for u, v in connections:
            uf.union(u - 1, v - 1)

        result = []
        
        for query in queries:
            if query[0] == 1:
                x = query[1] - 1
                if uf.online[x]:
                    result.append(x + 1)
                else:
                    result.append(uf.get_min_online(x))
            elif query[0] == 2:
                x = query[1] - 1
                uf.set_offline(x)

        return result